Add slow worm epic (GSI-1126) (#146)
ghga-de/epic-docs: 74-slow-worm/technical_specification.md

# Hexkit Support for BSON UUIDs and Datetimes (Slow Worm)
**Epic Type:** Implementation Epic

Epic planning and implementation follow the
[Epic Planning and Marathon SOP](https://docs.ghga-dev.de/main/sops/sop001_epic_planning.html).

## Scope
### Outline:

> The aim of this epic is to stop serializing UUIDs and Datetimes as strings in `hexkit`,

Reviewer: ```suggestion
The aim of this epic is to stop serializing UUID and datetime objects as strings in `hexkit`
and store them s BSON objects instead,
```

as decided in the ADR "[UUID and Datetime Representation in MongoDB](https://github.com/ghga-de/adrs/blob/main/docs/adrs/adr011_uuids_datetimes_mongodb.md)".

> The `hexkit` changes for this epic should be released with the Eurasian Blackbird

Reviewer: link 
```suggestion
The `hexkit` changes for this epic should be released with the [Eurasian Blackbird](../75-eurasian-blackbird/technical_specification.md)
```

Author: Strings are stored as BSON, too, but their structure is different.

`hexkit` changes. 

### Included/Required:

> - Update `hexkit` to remove string serialization for UUIDs and Datetimes, as well as

Reviewer: ```suggestion
- Update `hexkit` to remove string serialization for UUIDs and datetimes, as well as
```

  add support for binary UUIDs and tz-aware binary datetimes
- Rollout:

>   - Update `ghga-service-commons`, followed by `ghga-event-schemas`, then other services

Reviewer: ```suggestion
  - Update `ghga-service-commons`, followed by `ghga-event-schemas`, then other services:
```

    - Update to new `hexkit` version
    - Update references to `correlation_id` that are str-typed
    - Replace string UUIDs and Datetimes in tests and service code

>     - Write migrations for all services that store UUIDs and Datetimes as strings

Author: One of our prominent fields, `upload_date`, is actually a string. I think there was another instance or two, with the majority of changes being strings as UUIDs.

  
### Deployment Note:
The `NS`'s `notifications` collection should be dropped once deployment is complete.

## API Definitions:

### Payload Schemas for Events:

Models Containing String Datetimes or UUIDs in `ghga-event-schemas`:
- UploadDateModel
- FileUploadReceived
- FileUploadValidationSuccess
- FileUploadValidationFailure
- FileInternallyRegistered
- FileRegisteredForDownload
- NonStagedFileRequested
- FileStagedForDownload
- FileDownloadServed
- UserID
- User
- AccessRequestDetails
- UserIvaState

## Additional Implementation Details:

### Hexkit

Changes needed:

> - Remove UUID and Datetime-specific string serialization in the MongoDB provider

Reviewer: ```suggestion
- Remove UUID and datetime specific string serialization in the MongoDB provider
```
